src/coreclr/vm/pendingload.h

```diff
@@ -17,248 +17,404 @@
+        // Entry allocations were shifted to about 11 during that startup phase.
+        static constexpr int PreallocatedEntryCount = 2;
+
+private:
+        Shard() : m_shardCrst(CrstUnresolvedClassLock)
+        {}
 
-    TableEntry     **m_pBuckets;    // Pointer to first entry for each bucket
-    DWORD           m_dwNumBuckets;
+        Entry *m_pLinkedListOfActiveEntries = NULL;
+        Crst                  m_shardCrst;
+        Entry  m_preAllocatedEntries[PreallocatedEntryCount];
 
 public:
+        Crst* GetCrst()
+        {
+            LIMITED_METHOD_CONTRACT;
+            return &m_shardCrst;
+        }
+
+        Entry* FindPendingTypeLoadEntry(DWORD hash, const TypeKey& typeKey)
+        {
+            WRAPPER_NO_CONTRACT;
+            for (PendingTypeLoadTable::Entry *current = m_pLinkedListOfActiveEntries; current != NULL; current = current->m_pNext)
+            {
+                if (current->m_dwHash != hash)
+                    continue;
+                TypeKey entryTypeKey = current->GetTypeKey();
+                if (TypeKey::Equals(&typeKey, &entryTypeKey))
+                {
+                    return current;
+                }
+            }
+
+            return NULL;
+        }
+
+        void RemovePendingTypeLoadEntry(Entry* pEntry)
+        {
+            LIMITED_METHOD_CONTRACT;
+
+            Entry **pCurrent = &m_pLinkedListOfActiveEntries;
+
+            while (*pCurrent != pEntry)
+            {
+                pCurrent = &((*pCurrent)->m_pNext);
+            }
+            *pCurrent = (*pCurrent)->m_pNext;
+        }
+
+        Entry* InsertPendingTypeLoadEntry(DWORD hash, const TypeKey& typeKey, TypeHandle typeHnd)
+        {
+            STANDARD_VM_CONTRACT;
+            Entry* result = NULL;
+
+            for (int iEntry = 0; iEntry < PreallocatedEntryCount; iEntry++)
+            {
+                if (m_preAllocatedEntries[iEntry].IsUnused())
+                {
+                    result = &m_preAllocatedEntries[iEntry];
+                    result->SetTypeKey(typeKey);
+                    result->Init(m_pLinkedListOfActiveEntries, hash, typeHnd);
+                    break;
+                }
+            }
+            if (result == NULL)
+                result = new Entry(m_pLinkedListOfActiveEntries, hash, typeKey, typeHnd);
+            
+            m_pLinkedListOfActiveEntries = result;
+            return result;
+        }
 
 #ifdef _DEBUG
-    DWORD           m_dwDebugMemory;
+        void Dump();
 #endif
+    };
 
-    static PendingTypeLoadTable *Create(LoaderHeap *pHeap, DWORD dwNumBuckets, AllocMemTracker *pamTracker);
+    // This number chosen by experimentation with a fairly complex ASP.NET application that would naturally use about 40,000 Entry structures on startup.
+    // Entry allocations were shifted to about 11 during that startup phase.
```

Comment:
How does 11 relate to 31?

Reply:
Ah, its an unclear comment... we previously had needed to allocate and free 40,000 or so different Entry structs, and with the new code in that environment we needed to allocate around 11. (The exact number fluctuated based on which threads made progress in what order, so it wasn't a completely stable number, but that's expected.)